### PR TITLE
perf(event_dispatcher): lock-free emit via atomic shared_ptr snapshot

### PR DESCRIPTION
**Pull request by tkhquang**
## Summary

Replaces `shared_mutex`-guarded `emit()` with an `std::atomic<std::shared_ptr<const HandlerList>>` copy-on-write snapshot, and adds an atomic handler-count fast path for the zero-subscriber case. `emit()` and `emit_safe()` are now lock-free on the hot path; `subscribe()` / `unsubscribe()` serialize writers on a small `std::mutex` and publish a new immutable snapshot.

## Perf (mingw-release, median of 5 runs per side)

| Scenario | Subs | Before | After | Delta | Verdict |
|---|---:|---:|---:|---:|:---|
| `emit` | 0 | 103.9 ns | **6.0 ns** | **-94.2% (17x)** | REAL |
| `emit` | 1 | 120.1 ns | 94.4 ns | -21.4% | REAL |
| `emit` | 8 | 245.6 ns | 216.3 ns | -11.9% | REAL |
| `emit` | 64 | 1103.5 ns | 1092.1 ns | -1.0% | NOISE |
| `emit_concurrent_4_threads` (8 subs) | — | 517.9 ns | **248.2 ns** | **-52.1% (2.1x)** | REAL |
| `subscribe_unsub_roundtrip` | — | 446.0 ns | 1150.4 ns | +158.0% | REAL (expected) |

Variance-backed. An earlier single-run measurement suggested an 18% regression at 64 subscribers; across 5 runs per side the delta drops to within noise. Full table, per-run TSVs, and methodology in `docs/analysis/event_dispatcher_bench_v3.1.0/README.md`.

## Changes

- `include/DetourModKit/event_dispatcher.hpp`: atomic snapshot + handler counter; fast path in `emit`/`emit_safe`; writer mutex for COW mutations.
- `tests/test_event_dispatcher.cpp`: 3 new tests (`EmptyFastPath_SkipsLock`, `SnapshotStability_DuringEmit`, `SnapshotReclamation_NoLeak`) via `DMK_EVENT_DISPATCHER_INTERNAL_TESTING`.
- `tests/bench_event_dispatcher.cpp` + CMake wiring: opt-in `DMK_BUILD_BENCHMARKS=ON` builds a standalone `DetourModKit_bench` (no gtest dep).
- `src/hook_manager.cpp` + `tests/test_hook_manager.cpp`: loader-lock destructor now heap-allocates the leaked maps directly to avoid container copy fallbacks on move-only `VmtHookEntry`; `static_assert` guards the ownership contract.
- `.github/workflows/coverage-pages.yml`: parallel `msvc-verify` job; pages deploy now requires both MinGW coverage and MSVC verify to pass.
- Docs: `README.md`, `AGENTS.md`, `docs/tests/README.md` updated for the new model + bench harness.

## Test plan

- [x] 939 / 939 tests pass on `mingw-debug`.
- [x] `EventDispatcherTest.*` 29 / 29 pass, including the three new tests.
- [x] `DetourModKit_bench` builds and runs clean on `mingw-release`.
- [x] 5-run variance analysis confirms no regression on any emit row.
- [ ] MSVC build green via the new CI job.
- [ ] ASan / TSan run (optional follow-up).

## Summary by CodeRabbit

* **New Features**
  - Added `DMK_BUILD_BENCHMARKS` build option for standalone performance testing

* **Bug Fixes**
  - Improved hook manager destructor safety during critical system teardown

* **Documentation**
  - Updated EventDispatcher architecture and performance characteristics
  - Added benchmark results and test coverage documentation
  - Refined design guidelines for edge-case handling

### CODE REVIEW

File: tests/bench_event_dispatcher.cpp [hunk 1]
```diff
@@ -0,0 +1,251 @@
...
+    using Clock = std::chrono::steady_clock;
+
+    // Compiler barrier / value sink. Prevents the optimizer from noticing
+    // that the handler result is unused and deleting the whole emit loop.
+    volatile std::uint64_t g_sink = 0;
+
+    void noop_handler(const BenchEvent &e) noexcept
+    {
+        g_sink = g_sink + static_cast<std::uint64_t>(e.value);
+    }
```
[the remark this answers is not in the document]
Comment (tkhquang) [reply tests/bench_event_dispatcher.cpp]: @coderabbitai

**F1 (restore instruction in bench docs).** Fixed. `docs/analysis/event_dispatcher_bench_v3.1.0/README.md` now uses `git checkout main -- include/DetourModKit/event_dispatcher.hpp` (with a note that any pre-change commit also works) so the "before" run is actually against the baseline.

**F2 (noexcept writers that allocate).** Fixed. `unsubscribe()` and `clear()` now wrap the `std::make_shared` / `reserve` / `push_back` chain in `try/catch`. On `bad_alloc`:
- `unsubscribe()` leaves `handlers_` and `handler_count_` unchanged and returns `false`. `Subscription::reset()` already treats `false` as "defer retry via the stored lambda," so OOM rides on the existing reentrancy retry path with zero caller change.
- `clear()` is a best-effort teardown no-op. Keeps the `noexcept` contract intact.

OOM contract documented in both method Doxygen blocks and referenced from the class-level thread-safety note.

**F3 (lock-free wording in header).** Fixed. Replaced the blanket "lock-free on the hot path" with:
- The zero-subscriber fast path is **wait-free** (single `memory_order_acquire` counter load).
- The snapshot load avoids any reader lock; it is genuinely lock-free on DWCAS-backed toolchains (libstdc++ x86_64) and may use an implementation-internal short-critical-section bit lock on toolchains without DWCAS (notably MSVC STL).

**F4 (lock-free wording in README).** Fixed. `README.md` Event Dispatcher bullets reworded to match F3; keeps the fast-path win headline without overclaiming toolchain-independent lock-freedom.

**F5 (`static_assert` on map move noexcept).** Fixed. Added immediately before the `new (std::nothrow) HookMap(...)` / `VmtHookMap(...)` expressions:

```cpp
static_assert(std::is_nothrow_move_constructible_v<HookMap>,
              "HookMap move ctor must be noexcept to keep the loader-lock leak path safe.");
static_assert(std::is_nothrow_move_constructible_v<VmtHookMap>,
              "VmtHookMap move ctor must be noexcept to keep the loader-lock leak path safe.");
